worker/caasunitinit/initializer.go

```diff
@@ -0,0 +1,211 @@
+// Copyright 2019 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package caasunitinit
+
+import (
+	"bytes"
+	"os"
+	"path/filepath"
+
+	"github.com/juju/errors"
+	"github.com/juju/utils"
+	"gopkg.in/juju/names.v3"
+	"gopkg.in/juju/worker.v1"
+	"gopkg.in/juju/worker.v1/catacomb"
+
+	"github.com/juju/juju/agent/tools"
+	"github.com/juju/juju/caas"
+	"github.com/juju/juju/caas/kubernetes/provider"
+	"github.com/juju/juju/caas/kubernetes/provider/exec"
+	cmdutil "github.com/juju/juju/cmd/jujud/util"
+	"github.com/juju/juju/worker/caasoperator"
+	"github.com/juju/juju/worker/uniter"
+)
+
+type unitInitializer struct {
+	catacomb catacomb.Catacomb
+
+	config  UnitInitializerConfig
+	unitTag names.UnitTag
+}
+
+type UnitInitializerConfig struct {
+	// Logger for the worker.
+	Logger Logger
+
+	// UnitProviderIDFunc returns the ProviderID for the given unit.
+	UnitProviderIDFunc func(unit names.UnitTag) (string, error)
+
+	// Paths provides CAAS operator paths.
+	Paths caasoperator.Paths
+
+	// OperatorInfo contains serving information such as Certs and PrivateKeys.
+	OperatorInfo caas.OperatorInfo
+
+	// ExecClient is used for initilizing units.
+	ExecClient exec.Executor
+
+	// WriteFile is used to write files to the local state.
+	WriteFile func(string, []byte, os.FileMode) error
+
+	// TempDir is used for creating a temporary directory.
+	TempDir func(string, string) (string, error)
+}
+
+// NewUnitInitializer returns a new worker start function to initilize the Unit.
+func NewUnitInitializer(config UnitInitializerConfig, unitTag names.UnitTag) func() (worker.Worker, error) {
```

Comment:
This doesn't need to be a worker. Just have it as a method which takes a cancel channel. That channel is set up in the other worker and closed as needed by the other worker.